Mathlib/Tactic/DepRewrite.lean

```diff
@@ -0,0 +1,526 @@
+  trace[traceClsCast] "casting{indentExpr e'}\nto expected type{indentExpr et}"
+  let ctx ← read
+  checkCastAllowed e' te' ctx.cfg.castMode
+
+  /- Try casting from the inferred type (x ↦ p),
+  and to the expected type (p ↦ x).
+  In certain cases we need to cast in both directions (see `bool_dep_test`). -/
+  match ← castBack? e' te' ctx.x ctx.h with
+  | some e'' =>
+    let te'' ← inferType e''
+    if ← withAtLeastTransparency .default <| withNewMCtxDepth <| isDefEq te'' et then
+      trace[traceClsCast] "done with one cast (x ↦ p):{indentExpr e''}"
+      return e''
+
+    let motive ← mkLambdaFVars #[ctx.x, ctx.h] et
+    let e''' ← mkEqRec motive e'' ctx.h
+    trace[traceClsCast] "done with two casts (x ↦ p, p ↦ x):{indentExpr e'''}"
+    return e'''
+  | none =>
+    let motive ← mkLambdaFVars #[ctx.x, ctx.h] et
+    let e'' ← mkEqRec motive e' ctx.h
+    trace[traceClsCast] "done with one cast (x ↦ p):{indentExpr e''}"
+    return e''
+
+/-- Like `visitAndCast`, but does not insert casts at the top level.
+The expected types of certain subterms are computed from `et?`. -/
+-- TODO(WN): further speedup might come from returning whether anything
+-- was rewritten inside a `visit`,
+-- and then skipping the type correctness check if it wasn't.
+partial def visit (e : Expr) (et? : Option Expr) : M Expr :=
+  withTraceNode traceClsVisit (fun
+    | .ok e' => pure m!"{e} => {e'} (et: {et?})"
+    | .error _ => pure m!"{e} => 💥️") do
+  checkCache { val := e : ExprStructEq } fun _ => Meta.withIncRecDepth do
+  let ctx ← read
+  if e.hasLooseBVars then
+    throwError "internal error: forgot to instantiate"
+  if e.toHeadIndex == ctx.pHeadIdx && e.headNumArgs == ctx.pNumArgs then
+    -- We save the metavariable context here,
+    -- so that it can be rolled back unless `occs.contains i`.
+    let mctx ← getMCtx
+    -- Note that the pattern `ctx.p` is created in the outer lctx,
+    -- so bvars from the visited term will not be unified into the pattern.
+    if ← withTransparency ctx.cfg.transparency <| isDefEq e ctx.p then
+      let i ← modifyGet fun i => (i, i+1)
+      if ctx.cfg.occs.contains i then
+        return ctx.x
+      else
+        -- Revert the metavariable context,
+        -- so that other matches are still possible.
+        setMCtx mctx
+  match e with
+  | .mdata d b => return .mdata d (← visitAndCast b et?)
+  | .app f a =>
+    let fup ← visit f none
+    let tfup ← inferType fup
+    withAtLeastTransparency .default <| forallBoundedTelescope tfup (some 1) fun xs _ => do
+      let #[r] := xs |
+        throwError m!"term in function position was rewritten to a non-function{indentExpr fup}"
+      let tr ← inferType r
+      let aup ← visitAndCast a tr
+      return .app fup aup
+  | .proj n i b =>
+    let bup ← visit b none
+    let tbup ← inferType bup
+    if !tbup.isAppOf n then
+      throwError m!"projection type mismatch{indentExpr <| .proj n i bup}"
+    return .proj n i bup
+  | .letE n t v b bi =>
```

Comment:
@kmill I have fixed [your counterexample](https://github.com/leanprover-community/mathlib4/pull/23930/files#r2148828694) using an idea from @plp127. Unfortunately it generates quite ugly terms, so it hidden behind the `+letAbs` option; I was not able to find a better solution, and suggestions are welcome. Overall, though, I think we really can aim for completeness here.

Reply:
We now produce better terms for `let`s.